internal/agent/device/applications/controller.go

```diff
@@ -35,14 +36,29 @@ func (c *Controller) Sync(ctx context.Context, current, desired *v1alpha1.Device
 	c.log.Debug("Syncing device applications")
 	defer c.log.Debug("Finished syncing device applications")
 
-	currentAppProviders, err := provider.FromDeviceSpec(ctx, c.log, c.podman, c.readWriter, current)
+	currentAppProviders, err := provider.FromDeviceSpec(
+		ctx,
+		c.log,
+		c.podman,
+		c.readWriter,
+		current,
+		provider.WithVerify(),
```

Comment:
note: `WithVerify` ensures that providers are validated until after the images are pulled.